diff --git a/setup.py b/setup.py
--- a/setup.py
+++ b/setup.py
@@ -256,3 +256,3 @@ def get_extension_deep_ep_cpp():
     sources = ['csrc/deep_ep.cpp', 'csrc/kernels/runtime.cu', 'csrc/kernels/layout.cu', 'csrc/kernels/intranode.cu']
-    include_dirs = ['csrc/']
+    include_dirs = ['csrc/', '/usr/local/cuda/include/cccl/']
     library_dirs = []
@@ -363,3 +363,3 @@ if __name__ == '__main__':
         install_requires=[
-            'pynvml',
+            'nvidia-ml-py>=12.0.0',
         ],
